File: blender_extension/roblox_interface/roblox_server/receive_messages.py
import time, threading, copy, sys

def register(utils):
    process_formats = utils.import_module("process_formats")

    class ReceiveMessagesThread(threading.Thread):
        def __pause(self):
            self.paused = True
            while self.paused:
                time.sleep(0.001)
                if self.stop_thread:
                    sys.exit()

        def __unpause(self):
            self.paused = False
        
        def receive_message(self, new_buffer):
            self.buffers_queued.append(new_buffer)
            self.__unpause()

        def __read_buffer(self, data_size):
            read_buffer = self.buffer
            read_offset = self.offset
            while self.buf_len < self.offset + data_size:
                if len(self.buffers_queued) == 0:
                    self.__pause()
                self.buffer = self.buffers_queued.pop(0)
                self.offset -= self.buf_len
                self.buf_len = len(self.buffer)
                read_buffer += self.buffer

            self.offset += data_size
            return read_buffer, read_offset

        def __parse(self, format_data, **masks):
            parsed_type = type(format_data)
            if parsed_type is list:
                args = ()
                for sub_format_data in format_data:
                    args += self.__parse(sub_format_data, **masks)
                return args
            else:
                return process_formats.read_funcs[format_data["type"]](self, format_data, **masks)
        
        def stop(self):
            self.stop_thread = True

        def run(self):
            self.buffers_queued = []
            self.buffer = bytearray()
            self.buf_len = 0
            self.offset = 0
            self.stop_thread = False

            while not self.stop_thread:
                message_id, = process_formats.message_id_format.unpack_from(*self.__read_buffer(process_formats.message_id_format.size))
                args = self.__parse(process_formats.message_formats[message_id])
                for callback in process_formats.message_listeners[message_id]:
                    thread = threading.Thread(target=callback, args=copy.deepcopy(args))
                    thread.start()

    receiveThread = ReceiveMessagesThread(name = "blender_roblox_sync Receive Messages Thread")
    global receive_message
    def receive_message(*args):
        receiveThread.receive_message(*args)

    return {
        "threads": (receiveThread,)
    }

File: roblox_plugin/src/Server/receiveMessages.ts
import * as ProcessFormats from "./processFormats";
import PauseThread from "./pauseThread";

export interface ReceiveMessagesThread extends PauseThread {
	buffer: buffer;
	bufLen: number;
	offset: number;
	buffersQueued: buffer[];
}
export namespace ReceiveMessagesThread {
	export function receiveMessage(receiveThread: ReceiveMessagesThread, newBuffer: buffer) {
		receiveThread.buffersQueued.push(newBuffer);
		PauseThread.unpause(receiveThread);
	}
	export function readBuffer(receiveThread: ReceiveMessagesThread, dataSize: number): LuaTuple<[buffer, number]> {
		let readBuffer = receiveThread.buffer;
		let readBufferSize = buffer.len(readBuffer);
		const readOffset = receiveThread.offset;
		while (receiveThread.bufLen < receiveThread.offset + dataSize) {
			if (receiveThread.buffersQueued.size() === 0) {
				PauseThread.pause(receiveThread);
			}
			receiveThread.buffer = receiveThread.buffersQueued.shift()!;
			receiveThread.offset -= receiveThread.bufLen;
			receiveThread.bufLen = buffer.len(receiveThread.buffer);

			const newReadBufferSize = readBufferSize + receiveThread.bufLen;
			const newReadBuffer = buffer.create(newReadBufferSize);
			buffer.copy(newReadBuffer, 0, readBuffer);
			buffer.copy(newReadBuffer, readBufferSize, receiveThread.buffer);
			readBuffer = newReadBuffer;
			readBufferSize = newReadBufferSize;
		}

		receiveThread.offset += dataSize;
		return $tuple(readBuffer, readOffset);
	}

	export function parse(
		receiveThread: ReceiveMessagesThread,
		args: defined[],
		formatData: ProcessFormats.FormatData,
		masks: Map<string, boolean>,
	) {
		if (!("type" in formatData)) {
			for (const subFormatData of formatData) {
				parse(receiveThread, args, subFormatData, masks);
			}
		} else {
			ProcessFormats.readFuncs.get(formatData.type)!(receiveThread, args, formatData, masks);
		}
	}

	export function create(): ReceiveMessagesThread {
		const receiveThread: ReceiveMessagesThread = {
			buffer: buffer.create(0),
			bufLen: 0,
			offset: 0,
			buffersQueued: [],
			stopThread: false,
		};

		task.spawn(() => {
			while (!receiveThread.stopThread) {
				const format = ProcessFormats.messageIdFormat;
				const [readBuffer, readOffset] = ReceiveMessagesThread.readBuffer(receiveThread, format.size);
				const messageId = format.read(readBuffer, readOffset);
				const args: defined[] = [];
				const masks = new Map<string, boolean>();
				parse(receiveThread, args, ProcessFormats.messageFormats[messageId], masks);
				for (const callback of ProcessFormats.messageListeners[messageId]) {
					task.spawn(callback, ...ProcessFormats.deepcopy(args));
				}
			}
		});
		return receiveThread;
	}
}
export default ReceiveMessagesThread;
